arangod/Cluster/Maintenance.cpp

```diff
@@ -1352,12 +1403,12 @@ static std::tuple<VPackBuilder, bool, bool> assembleLocalCollectionInfo(
               status->status.activeParticipantsConfig->participants.size();
           shardReplicated = !status->status.followersWithSnapshot.empty();
         } else {
-          ret.add(VPackValue(maintenance::SERVERS));
-          ret.add(VPackSlice::emptyArraySlice());
-          ret.add(VPackValue(StaticStrings::FailoverCandidates));
-          ret.add(VPackSlice::emptyArraySlice());
-          shardInSync = false;
-          shardReplicated = false;
+          THROW_ARANGO_EXCEPTION_MESSAGE(
```

Comment:
If the status cannot be retrieved successfully, the replicated log is not the leader (or does not even exist anymore, but that's improbable) - in most cases, it is a follower.
This happens rarely, but note that in the previous version, the follower wrote a bunch of empty arrays to `Current`. That comes with two bad consequences:
- The follower writes freely to `Current`, potentially overwriting the leader (was erlaube)
- The values overwritten are actually needed in `ClusterInfo::getResponsibleServer` and used for resolving the destination of cluster requests. This opens the door for a race where in `network::resolveDestination` we temporarily get `CLUSTER_BACKEND_UNAVAILABLE`. I've seen tests failing spuriously due to this.

My changes make sure we only report in `Current` from a server that is leader, see `Maintenance.cpp:1980`:
```
if (replicationVersion == replication::Version::TWO && !isReplication2Leader(shName, logs->second, shardsToLogs->second)) {
    // skip this shard if it is not the leader
    continue;
}
```
In theory, we should never get to this situation again. For maintainer mode, note the assertion in `Maintenance.cpp:1437` (right before the `if (status)`:
```
 TRI_ASSERT(status != nullptr)
```
The `THROW_ARANGO_EXCEPTION_MESSAGE` is just a way of signaling that something went really wrong, if we're in production and the assertion is not compiled. I deliberately chose to throw an exception because the entire code block is wrapped in a `try/catch` which does proper exception handling, so it's certain that the exception is not going to bubble up.

Reply:
Happy to go through the code together for further explanation

Reply:
No need to go through the code, looks good :+1: Good change